tensorflow/python/ops/special_math_ops.py

```diff
@@ -452,6 +485,188 @@ def _total_size(shape_values):
   return result
 
 
+
+def _einsum_optimize_dp(ishapes, ilabels, olabels, cost_limit=np.inf):
+  # decompose the contraction graph into connected subgraphs and optimise
+  # each subgraph using _einsum_optimize_dp_connected
+  c = [
+    _einsum_optimize_dp_connected(
+      [ishapes[j] for j in g], 
+      [ilabels[j] for j in g], 
+      olabels, cost_limit, g
+    ) for g in _find_subgraphs(ilabels, olabels)
+  ]
+  return _tree_to_sequence(reduce(lambda c1, c2: (c1, c2), c))
+
+
+def _find_subgraphs(ilabels, olabels):
+  subgraphs = []
+  unused = set(range(len(ilabels)))
+  
+  while len(unused) > 0:
+    g = []
+    q = [unused.pop()]
+    while len(q) > 0:
+      x = q.pop(0)
+      g.append(x)
+      n = {
+        y for y in unused 
+        if len((set(ilabels[x]) & set(ilabels[y])) - set(olabels)) > 0
+      }
+      q.extend(n)
+      unused -= n
+    
+    subgraphs.append(g)
+    
+  return subgraphs
+
+
+def _einsum_optimize_dp_connected(ishapes, ilabels, olabels, 
+                                  cost_limit=np.inf, tensor_indices=None):
+                                  
+  # find an optimal contraction using breadth-first search with dynamic 
+  # programming but ignoring solutions with intermediate outer products
+  # and solutions with contraction cost larger than cost_limit
+  
+  n = len(ishapes)
+  if tensor_indices is None: 
+    tensor_indices = list(range(n))
+  x = [
+    None, # just ignore x[0]
+    {
+      frozenset([j]): (ishapes[j], ilabels[j], 0, tensor_indices[j]) 
+      for j in range(n)
+    }
+  ]
+  # x[n_tensors][set of tensors] = (shape, labels, cost, contraction)
+  
+  for m in range(2, n+1): # construct x[m]
+    x.append(dict())
+    
+    for k in range(1, m//2+1): # try to combine all x[m-k] and x[k]
+    
+      for s1 in x[m-k]:
+        d1, l1, c1, e1 = x[m-k][s1]
+        
+        for s2 in x[k]:
+          if len(s1 & s2) == 0:
+            d2, l2, c2, e2 = x[k][s2]
+            
+            s = s1 | s2
+            
+            common_indices = set(l1) & set(l2)
+            contraction_indices = common_indices - set(olabels)
+            
+            if len(contraction_indices) > 0: # ignore outer products
+
+              # m1[l] <-- is l1[l] an index of the new tensor? (m2 for l2)
+              m1 = [l not in contraction_indices for l in l1]
+              m2 = [l not in common_indices      for l in l2]
+
+              new_shape = tuple(
+                [d for d, f in zip(d1, m1) if f] + \
+                [d for d, f in zip(d2, m2) if f]
+              )
+
+              new_labels = \
+                ''.join([l for l, f in zip(l1, m1) if f]) + \
+                ''.join([l for l, f in zip(l2, m2) if f])
+              
+              contraction_cost = np.prod(d1) * \
```

Comment:
What happens if one of the input shapes is unknown? 
Maybe I missed it, but I don't see a any check for [None] in the shapes.
Don't you need to fallback to "optimize=False" in that case?

Reply:
Already now `einsum` just fails, if input shapes are not given. Therefore, also `optimize=False` won't work. Just try the following code:
```python
a = tf.placeholder(tf.float64)
b = tf.placeholder(tf.float64)
c = tf.einsum("ij,jk", a, b)
```
This results with:
```
ValueError: Cannot take the length of Shape with unknown rank.
```

Reply:
I wouldn't expect that to work either.

The case I'm interested in is with known ranks, but unknown dimensions:

```
a = tf.placeholder(tf.float64, [10,None])
b = tf.placeholder(tf.float64, [None,7])
c = tf.einsum("ij,jk", a, b)
```
```
<tf.Tensor 'einsum_2/transpose_2:0' shape=(10, 7) dtype=float64>
```

For a real world example think of the transformer attention layers, where the sequence lengths are unknown at "compile" time. 

```
# query is [batch, sequence_s, depth]
# key is [batch, sequence_t, depth]

einsum('bsd,btd-->bst', query, key)
```